File: service/simple-python/seller.py
#!/usr/bin/env python3
"""Simple Seller example.
"""

from autobahn.twisted.xbr import SimpleSeller
from autobahn.twisted.util import sleep
from autobahn.twisted.component import Component, run
from autobahn.wamp.types import PublishOptions
import binascii
import os
from time import sleep
from uuid import UUID
import txaio
txaio.use_twisted()


DEFAULT_AUTHID = 'team1'
DEFAULT_TICKET = 'apple pear orange fig'

comp = Component(
    transports=os.environ.get(
        'XBR_INSTANCE', 'wss://continental2.crossbario.com/ws'),
    realm=os.environ.get('XBR_REALM', 'realm1'),
    authentication={
        'ticket': {
            'authid': os.environ.get('XBR_AUTHID', DEFAULT_AUTHID),
            'ticket': os.environ.get('XBR_TICKET', DEFAULT_TICKET),
        },
    },
    extra={
        'market_maker_adr': os.environ.get('XBR_MARKET_MAKER_ADR', '0x3e5e9111ae8eb78fe1cc3bb8915d5d461f3ef9a9'),
        'seller_privkey': os.environ.get('XBR_SELLER_PRIVKEY', '0x496fbc218ffb1afe839b93cce01819433df1b72034a94ec46d83592c7342d632'),
    }
)

running = False


@comp.on_join
async def joined(session, details):
    print('Seller session joined', details)
    global running
    running = True

    market_maker_adr = binascii.a2b_hex(
        session.config.extra['market_maker_adr'][2:])
    print('Using market maker adr:', session.config.extra['market_maker_adr'])

    seller_privkey = binascii.a2b_hex(
        session.config.extra['seller_privkey'][2:])

    api_id = UUID('627f1b5c-58c2-43b1-8422-a34f7d3f5a04').bytes
    topic = 'com.conti.hackathon.team1.sell'
    counter = 1

    seller = SimpleSeller(market_maker_adr, seller_privkey)
    price = 35 * 10 ** 18
    interval = 10
    seller.add(api_id, topic, price, interval, None)

    balance = await seller.start(session)
    balance = int(balance / 10 ** 18)
    print("Remaining balance: {} XBR".format(balance))

    while running:
        payload = {'data': 'py-seller', 'counter': counter}
        key_id, enc_ser, ciphertext = await seller.wrap(api_id,
                                                        topic,
                                                        payload)
        pub = await session.publish(topic, key_id, enc_ser, ciphertext,
                                    options=PublishOptions(acknowledge=True))

        print('Published event {}: {}'.format(pub.id, payload))

        counter += 1
        await sleep(10)


@comp.on_leave
def left(session, details):
    print('Seller session left', details)
    global running
    running = False


if __name__ == '__main__':
    run([comp])

File: service/simple-js/seller.js
var autobahn = require('autobahn');
var xbr = require('autobahn-xbr');
var web3 = require('web3');
var BN = web3.utils.BN;
var uuid = require('uuid-buffer');

console.log('Running on Autobahn ' + autobahn.version);
//Private key of your seller delegate
const SELLER_PRIVATEKEY = "0xA546A4953D3953EB3A6BE8E64364936F71D1C51913A3A02180D15CEA352B2A0D";
//Market maker address
const MARKET_MAKER_ADDR = "0x3E5e9111Ae8eB78Fe1CC3bb8915d5D461F3Ef9A9";
//TEAM_ID
const TEAM_ID = "team1";
//TICKET
const TICKET = "apple pear orange fig";
//TOPIC TO SUBSCRIBE TO
const TOPIC = "com.conti.hackathon.team1.topic1";

//Connect to Crossbarfx node
var connection = new autobahn.Connection({
    realm: "realm1",
    authmethods: ["ticket"],
    authid: TEAM_ID,
    onchallenge: onchallenge,
    transports: [
      {
        url: "wss://continental2.crossbario.com/ws",
        type: "websocket",
        serializers: [new autobahn.serializer.CBORSerializer()]
      }
    ]
  });
  
  //Only for authentication on crossbar
  function onchallenge(session, method, extra) {
    if (method === "ticket") {
      return TICKET;
    } else {
      throw "don't know how to authenticate using '" + method + "'";
    }
  };

// callback fired upon new WAMP session
connection.onopen = function (session, details) {

    console.log("WAMP session connected:", details);

    // the XBR token has 18 decimals
    const decimals = new BN('1000000000000000000');

    // price in XBR per key
    const price = new BN(35).mul(decimals);

    // key rotation interval in seconds
    const key_rotation_interval = 10;

    // API ID of the interface of the offered service
    const apiID = uuid.toBuffer('bd65e220-aef7-43c0-a801-f8d95fa71f39');

    // instantiate a simple seller
    var seller = new xbr.SimpleSeller(MARKET_MAKER_ADDR, SELLER_PRIVATEKEY);
    var counter = 1;

    seller.add(apiID, TOPIC, price, key_rotation_interval);

    var do_publish = function() {
        const payload = {"data": "js-seller", "counter": counter};

        // encrypt the XBR payload, potentially automatically rotating & offering a new data encryption key
        let [key_id, enc_ser, ciphertext] = seller.wrap(apiID, TOPIC, payload);

        const options = {acknowledge: true};

        session.publish(TOPIC,
                        [key_id, enc_ser, ciphertext],
                        {},
                        options).then(
            function (pub) {
                console.log("Published event " + pub.id);
            },
            function (err) {
                console.log("Failed to publish event", err);
            }
        );

        counter += 1;
        setTimeout(do_publish, 1000);
    };

    // start selling
    seller.start(session).then(
        // success: we've got an active paying channel with remaining balance ..
        function (balance) {
            console.log("Seller has started, remaining balance in active paying channel is " + balance.div(decimals) + " XBR");
            do_publish();
        },
        // we don't have an active paying channel => bail out
        function (error) {
            console.log("Failed to start seller:", error);
            process.exit(1);
        }
    );
};

// open WAMP session
connection.open();
